src/ngraph/dimension.hpp

```diff
@@ -0,0 +1,84 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <limits>
+#include <stddef.h>
+
+namespace ngraph
+{
+    /// \brief Class representing a possibly-unknown dimension in a shape or shape-like object.
+    ///
+    ///        Known dimensions may be implicitly converted from size_t. An unknown dimension is
+    ///        constructed with Dimension() or Dimension::undetermined().
+    ///
+    /// XXX: THIS CLASS IS NOT IN USE YET AND THE ENTIRE DESIGN IS SUBJECT TO CHANGE.
+    class Dimension
+    {
+    public:
+        /// \brief Constructs a known dimension.
+        ///
+        ///        Requires that size_t != std::numeric_limits<size_t>::max(). If that condition
+        ///        does not hold, throws std::invalid_argument.
+        Dimension(size_t dimension)
+            : m_dimension(dimension)
+        {
+            if (dimension == s_undetermined_val)
+            {
+                throw std::invalid_argument(
+                    "Cannot convert std::numeric_limits<size_t>::max() to Dimension");
```

Comment:
Use that value of `s_undetermined_val` instead of `std::numeric_limits<size_t>::max()` and then it will always be the correct value. And the message should have a brief why in it.